### PR TITLE
[FLINK-28709] Implement dynamic filtering operators

### PR DESCRIPTION
**Pull request by pltbkd**
## What is the purpose of the change

This pull request introduces operators to support dynamic filtering. 

## Brief change log

  - Introduce coordinatorListeningID in SourceCoordinator to listen to events from other coordinators.
  - Introduce DynamicFilteringData and the DynamicFilteringDataCollectorOperator to build and distribute the data.
  - Introduce ExecutionOrderEnforcerOperator to ensure the source with dynamic filtering is executed after the DynamicFilteringData is collected.

## Verifying this change

This change added new tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorProvider.java [hunk 1]
```diff
@@ -82,12 +83,18 @@ public OperatorCoordinator getCoordinator(OperatorCoordinator.Context context) {
...
                 operatorName,
                 source,
                 sourceCoordinatorContext,
                 context.getCoordinatorStore(),
                 alignmentParams);
+        coordinator.setCoordinatorListeningID(coordinatorListeningID);
+        return coordinator;
+    }
+
+    public void setCoordinatorListeningID(String coordinatorListeningID) {
```
Comment (gaoyunhaii) [inline flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorProvider.java:97]: Might make `coordinatorListeningID` to be a constructor parameter?

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorProvider.java [hunk 2]
```diff
@@ -43,6 +43,7 @@ Licensed to the Apache Software Foundation (ASF) under one
     private final Source<?, SplitT, ?> source;
     private final int numWorkerThreads;
     private final WatermarkAlignmentParams alignmentParams;
+    private String coordinatorListeningID;
```
Comment (gaoyunhaii) [inline flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorProvider.java:46]: `@Nullable`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+    public DynamicFilteringData(
+            TypeInformation<RowData> typeInfo,
+            RowType rowType,
+            List<byte[]> serializedData,
+            boolean isFiltering) {
+        this.typeInfo = typeInfo;
+        this.rowType = rowType;
+        this.isFiltering = isFiltering;
+        this.serializedData = isFiltering ? serializedData : Collections.emptyList();
```
Comment (gaoyunhaii) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:60]: Is it necessary to do the distinguish ?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+    public boolean isFiltering() {
+        return isFiltering;
+    }
+
+    public RowType getRowType() {
+        return rowType;
+    }
+
+    public Collection<RowData> getData() {
```
Comment (gaoyunhaii) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:71]: `@VisibleForTesting` ?

And the two method might be moved to the end of the file.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            RowType rowType,
+            List<byte[]> serializedData,
+            boolean isFiltering) {
+        this.typeInfo = typeInfo;
+        this.rowType = rowType;
+        this.isFiltering = isFiltering;
+        this.serializedData = isFiltering ? serializedData : Collections.emptyList();
+    }
+
+    public boolean isFiltering() {
```
Comment (gaoyunhaii) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:63]: `@VisibleForTesting` ?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        return isFiltering;
+    }
+
+    public RowType getRowType() {
```
Comment (gaoyunhaii) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:67]: This method is not used now, does it necessary in the future ?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    public boolean contains(RowData row) {
+        if (!isFiltering) {
+            return true;
+        } else if (row.getArity() != rowType.getFieldCount()) {
+            throw new TableException("The arity of RowData is different");
+        } else {
+            prepare();
+            RowData rowData = dataMap.get(hash(row));
```
Comment (gaoyunhaii) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:83]: Would the hash conflicts in some cases?

If so, we might lost some partitions?

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    private transient Set<byte[]> buffer;
+    private transient long currentSize;
+    private transient boolean eventSent;
+
+    public DynamicFilteringDataCollectorOperator(
+            RowType dynamicFilteringFieldType,
+            List<Integer> dynamicFilteringFieldIndices,
+            long threshold,
+            OperatorEventGateway operatorEventGateway) {
+        this.dynamicFilteringFieldType = dynamicFilteringFieldType;
```
Comment (gaoyunhaii) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java:75]: nit: add checkNotNull when possible

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+                    rowData.setField(i, value.getInt(index));
+                    break;
+                case BIGINT:
+                    rowData.setField(i, value.getLong(index));
+                    break;
+                case VARCHAR:
+                    rowData.setField(i, value.getString(index));
+                    break;
+                default:
+                    throw new UnsupportedOperationException();
```
Comment (gaoyunhaii) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java:113]: nit: Add some message

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+            }
+        }
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
Comment (gaoyunhaii) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java:117]: wrapped with try-with-resource ?

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+        DataOutputViewStreamWrapper wrapper = new DataOutputViewStreamWrapper(baos);
+        serializer.serialize(rowData, wrapper);
+        boolean duplicated = !buffer.add(baos.toByteArray());
```
Comment (gaoyunhaii) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java:120]: Why need this line? The byte array seems always newly-created, and byte[] should not override the hashcode / equals method.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+        if (duplicated) {
+            return;
+        }
+        currentSize += baos.size();
+
+        if (exceedThreshold()) {
+            // Send an empty filtering data and disable self by leaving the currentSize unchanged
+            sendEvent();
```
Comment (gaoyunhaii) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java:128]: This seems not necessary? Could we directly do it on finishing and remove the `eventSent` flag ?

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+            buffer.clear();
+            LOG.info(
```
Comment (gaoyunhaii) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java:130]: nit: change to warn?

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+    private void sendEvent() {
+        if (eventSent) {
+            return;
+        }
+
+        final DynamicFilteringData dynamicFilteringData;
+        if (exceedThreshold()) {
+            dynamicFilteringData =
+                    new DynamicFilteringData(
+                            typeInfo, dynamicFilteringFieldType, new ArrayList<>(), false);
```
Comment (gaoyunhaii) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java:161]: nit: Collections.emptyList()

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(DynamicFilteringDataCollectorOperatorCoordinator.class);
+
+    private final CoordinatorStore coordinatorStore;
+    private final List<String> dynamicFilteringDataListenerIDs;
+
+    private boolean hasReceivedFilteringData;
+
+    public DynamicFilteringDataCollectorOperatorCoordinator(
```
Comment (gaoyunhaii) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java:53]: nit: checkNotNull when possible

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+        // has been executed for multiple attempts, we only keep the first notification.
+        if (hasReceivedFilteringData) {
+            return;
+        }
+
+        for (String listenerID : dynamicFilteringDataListenerIDs) {
+            // Push event to listening source coordinators.
+            OperatorCoordinator listener = (OperatorCoordinator) coordinatorStore.get(listenerID);
+            if (listener == null) {
+                LOG.warn("Dynamic filtering data listener is missing: {}", listenerID);
```
Comment (gaoyunhaii) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java:78]: Would this be marked as an error?

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/ExecutionOrderEnforcerOperator.java [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * ExecutionOrderEnforcerOperator has two inputs, one of which is a source, and the other is the
+ * dependent upstream. It enforces that the input source is executed after the dependent input is
+ * finished. Everything passed from the inputs is forwarded to the output, though typically the
+ * dependent input should not send anything.
+ *
+ * <p>The operator must be chained with the source, which is ensured by the {@link
```
Comment (gaoyunhaii) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/ExecutionOrderEnforcerOperator.java:40]: This might not be very accurate since users have method to disable chaining. We might change it to say that if not chained, the enforcer would not work, or modify the last paragraph to say that it enforces... when chaining

**Review by gaoyunhaii — COMMENTED** [review]
Thanks @pltbkd for the PR! I have left some comments~

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java (hunk 6 above)
Comment (pltbkd) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:67]: Yes, this is necessary for building a row for comparing.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java (hunk 7 above)
Comment (pltbkd) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:83]: Here seems to be a bug. I'll fix it.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java (hunk 11 above)
Comment (pltbkd) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java:120]: buffer is a TreeSet with a byte[] comparator, so different byte[] instances with same content will be deduplicated as well.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java (hunk 12 above)
Comment (pltbkd) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java:128]: An empty event takes no effect to the source. It seems that we don't need to eagerly send the event. Though it's still better to clear the buffer eagerly. I'll remove the sending here, as well as the flag.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java (hunk 16 above)
Comment (pltbkd) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java:78]: Maybe it's better to be an error. The exception will not make the result incorrect, but the dynamic filtering will be invalid. So it's not a critical error, but indeed an error. I'll modify the log level.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java (hunk 5 above)
Comment (pltbkd) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:65]: It's also necessary in the future.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+
+    @Override
+    public void close() throws Exception {}
+
+    @Override
+    public void handleEventFromOperator(int subtask, int attemptNumber, OperatorEvent event)
+            throws Exception {
+        // Since there might be speculative execution, once the dynamic filter collectors operator
+        // has been executed for multiple attempts, we only keep the first notification.
+        if (hasReceivedFilteringData) {
```
Comment (gaoyunhaii) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java:72]: If there are global failover, all the tasks would restarted, then here we might have some problem since the flag  would not be clear. We might make the provider of this operator coordinator extends the `RecreateOnResetOperatorCoordinator.Provider`.

If the dpp task restarted, it's also possible that the filtering data get changed. In this case, we have to ensure the filtering data sent to the source coordinator and the join operators are same. However, in consideration of the speculative execution, we could not easily know which version is the chosen one. Thus for now we may need to ask user to disable dpp in this case. For the long run, we may consider restarting sources in some way once the data changed. 

In summary, we may
1. Use RecreateOnResetOperatorCoordinator to clear the state on global failover.
2. When received repeat data, compare it to the previous one, if not the same, then throw an exception now.
3. Create a follow-up issue to optimize this behavior (restarting sources or update partition is source is not fully finished) in the future.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+            return;
+        }
+
+        for (String listenerID : dynamicFilteringDataListenerIDs) {
+            // Push event to listening source coordinators.
+            OperatorCoordinator listener = (OperatorCoordinator) coordinatorStore.get(listenerID);
+            if (listener == null) {
+                LOG.error("Dynamic filtering data listener is missing: {}", listenerID);
```
Comment (gaoyunhaii) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java:80]: Not about the log level, perhaps we might directly throw an exception in this case?

Logically if a dpp operator is created, there must be at least one corresponding fact source.

Comment (pltbkd) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java]: Yes, it can't more than a bug indeed. I'll make it throw exception.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java (hunk 18 above)
Comment (pltbkd) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java]: Failover handling can be indeed complex for this coordinator and the receiver(in the next PR). I agree that we are ready to handle it gracefully. I'll follow the suggestion as it should be the safest way for now.

**Comment by pltbkd** [comment]
Hi @gaoyunhaii ,
Thanks for the reviewing! All comments so far should have been addressed and resolved.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java [hunk 20]
```diff
@@ -67,17 +69,32 @@ public void close() throws Exception {}
...
+            // Since there might be speculative execution or failover, we may receive multiple
+            // notifications, and we can't tell for sure which one is valid for further processing.
+            if (receivedFilteringData.equals(currentData)) {
+                // If the notifications contain exactly the same data, everything is alright, and
+                // we don't need to send the event again.
+                return;
+            } else {
+                // In case the mismatching of the source filtering result and the dim data, which
+                // may leads to incorrect result, trigger global failover for fully recomputing.
+                throw new IllegalStateException(
```
Comment (gaoyunhaii) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinator.java:87]: Might tip users to disable dpp in this case.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java [hunk 21]
```diff
@@ -141,6 +144,24 @@ private int hash(RowData row) {
         return Objects.hash(Arrays.stream(fieldGetters).map(g -> g.getFieldOrNull(row)).toArray());
     }
 
+    public boolean equals(DynamicFilteringData another) {
```
Comment (gaoyunhaii) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java]: Might change it to a static method outside of the DynamicFilteringData if we do not want to make it to be Object#equals

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+
+/** Data for dynamic filtering. */
+public class DynamicFilteringData implements Serializable {
+    private final TypeInformation<RowData> typeInfo;
+    private final RowType rowType;
+
+    /** The list should be sorted and distinct. */
+    private final List<byte[]> serializedData;
+
+    private final boolean isFiltering;
```
Comment (godfreyhe) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:53]: please  add some comments about this field

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+            if (comparator.compare(data.serializedData.get(i), another.serializedData.get(i))
+                    != 0) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @VisibleForTesting
+    public Collection<RowData> getData() {
```
Comment (godfreyhe) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:170]: this method is not only used for testing, but also can be used for advanced filtering which `contains` methods can not meet the requirement

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+                }
+            }
+        }
+    }
+
+    private int hash(RowData row) {
+        return Objects.hash(Arrays.stream(fieldGetters).map(g -> g.getFieldOrNull(row)).toArray());
+    }
+
+    public static boolean isEqual(DynamicFilteringData data, DynamicFilteringData another) {
```
Comment (godfreyhe) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:147]: implements `equals` method directly ?

**Comment by gaoyunhaii** [comment]
@flinkbot run azure

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+                case INTEGER:
+                    rowData.setField(i, value.getInt(index));
+                    break;
+                case BIGINT:
+                    rowData.setField(i, value.getLong(index));
+                    break;
+                case VARCHAR:
+                    rowData.setField(i, value.getString(index));
+                    break;
+                default:
```
Comment (godfreyhe) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java:113]: we should more types, such as: CHAR, SMALLINT, TINYINT, etc. It's better we can defined the supported types in `DynamicFilteringData` , and then DynamicPartitionPruningRule can skip the query which contains unsupported types.

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorTest.java [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Tests for {@link DynamicFilteringDataCollectorOperator}. */
+class DynamicFilteringDataCollectorOperatorTest {
```
Comment (godfreyhe) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorTest.java:45]: add some test for unsupported types

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinatorTest.java [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Test for {@link DynamicFilteringDataCollectorOperatorCoordinator}. */
+class DynamicFilteringDataCollectorOperatorCoordinatorTest {
```
Comment (godfreyhe) [inline flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorCoordinatorTest.java:40]: add some tests for failover case

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution @pltbkd , I left some comments

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java [hunk 28]
```diff
@@ -0,0 +1,183 @@
...
+                    "Finish collecting. {} bytes are collected which exceeds the threshold {}. Sending empty data.",
+                    currentSize,
+                    threshold);
+        } else {
+            LOG.info(
+                    "Finish collecting. {} bytes in {} rows are collected. Sending the data.",
+                    currentSize,
+                    buffer.size());
+        }
+        sendEvent();
```
Comment (godfreyhe) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java:157]: what if we do not send any event when `exceedThreshold` ?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java (hunk 24 above)
Comment (pltbkd) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:147]: As the data struct may contain large scale of data, calculate the hashcode with all the data can be costing. So I suppose we'd better not to override the hashcode, as well as the equals.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java (hunk 23 above)
Comment (pltbkd) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:170]: So far the method is not used in formal code, including the dynamic filtering implementation of hive source. I'd prefer to remain the annotation for now. We can remove it when it's actually used.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java (hunk 25 above)
Comment (pltbkd) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java:113]: I suppose we can add a static method giving all the supported types as a Set for this usage. maybe 
Set<LogicalTypeRoot> getSupportedTypes()
Besides, since the setters of GenericRowData don't care the type, I suppose we can create FieldGetters for getting value from the input rowType, and directly set to the GenericRowData we create here in spite of the type. The code can be much cleaner and we can support more types easily.

File: flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorTest.java (hunk 26 above)
Comment (pltbkd) [reply flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperatorTest.java:45]: With the changes above, the operator need no longer check for unsupported types. So I think we don't need the test now.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java (hunk 28 above)
Comment (pltbkd) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java:157]: In case different attempts of the collector calculate different results, one of which is exceeding threshold, I think it's better to send the event even it's exceeding threshold, or the coordinator may not know the problem and can't trigger failover for recomputing.

**Comment by pltbkd** [comment]
Hi @godfreyhe ,
Thanks for the suggestions!
I suppose I've resolved most of the comments, and the failover case is coming soon. Would you please take another look?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java [hunk 29]
```diff
@@ -26,30 +26,49 @@
...
+
+    static {
+        Set<LogicalTypeRoot> supported = new HashSet<>();
+        supported.add(LogicalTypeRoot.INTEGER);
+        supported.add(LogicalTypeRoot.BIGINT);
+        supported.add(LogicalTypeRoot.SMALLINT);
+        supported.add(LogicalTypeRoot.TINYINT);
+        supported.add(LogicalTypeRoot.VARCHAR);
+        supported.add(LogicalTypeRoot.CHAR);
+        SUPPORTED_TYPES = Collections.unmodifiableSet(supported);
```
Comment (godfreyhe) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:62]: I think date/time/timestamp can be supported

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java (hunk 23 above)
Comment (godfreyhe) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:189]: I think it should be a public interface, some lake house connectors has statistics info in partition spec, the partition can be skipped not only based on the partition value but also base on the statistics. If these connectors want to support dpp, they will not change the flink framework code.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java (hunk 29 above)
Comment (pltbkd) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:67]: DATE, TIME_WITHOUT_TIME_ZONE, TIMESTAMP_WITH_TIME_ZONE, TIMESTAMP_WITH_LOCAL_TIME_ZONE, TIMESTAMP_WITHOUT_TIME_ZONE are added. 
In fact as the field values are acquired and used to form a GenericRowData in Hive source in the same way as in the collector operator, I suppose all types are supported at runtime level as long as the type is exactly the same. The supported types set only depends on what types are supported at planner level.

Comment (pltbkd) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:67]: My apology.  I mistook the way hive source to acquire the data. It seems that we can support all types that hive table spec can describe (in string) and can be converted to the runtime type, but need to enumerate these types in Hive source.

Comment (pltbkd) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:67]: Seems that we can't support TIMESTAMP_WITH_TIME_ZONE at present. I'm removing it from the supported types.

**Comment by gaoyunhaii** [comment]
@flinkbot run azure

**Comment by pltbkd** [comment]
@flinkbot run azure

**Comment by godfreyhe** [comment]
@flinkbot run azure

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java [hunk 30]
```diff
@@ -0,0 +1,207 @@
...
+            }
+            return false;
+        }
+    }
+
+    private boolean matchRow(RowData row, RowData mayMatch) {
+        for (int i = 0; i < rowType.getFieldCount(); ++i) {
+            if (!Objects.equals(
+                    fieldGetters[i].getFieldOrNull(row),
+                    fieldGetters[i].getFieldOrNull(mayMatch))) {
```
Comment (godfreyhe) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:125]: for Date, Time, Timestamp, the inner representation is int/long, but external representation is java.sql.Date, java.sql.Time, java.sql.Timestamp. We should make sure they can be compared. Please also add some tests in DynamicFilteringDataTest

Comment (pltbkd) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:125]: Values in the building rows and testing rows are all expected to be Flink internal types, which should be guaranteed by the builders and users. I suppose that's not the responsibility of the DynamicFilteringData. But maybe I can add some comments to explain this to notify all users.

Comment (pltbkd) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/DynamicFilteringData.java:125]: In addition, here users mean the users of this data class, rather than flink users. So I suppose the behavior of the users  should be under control.

**Review by godfreyhe — APPROVED** [review]
Thanks for the update, LGTM

**Comment by gaoyunhaii** [comment]
@flinkbot run azure

**Comment by gaoyunhaii** [comment]
@flinkbot run azure
